barretenberg/cpp/src/barretenberg/boomerang_value_detection/graph.cpp

```diff
@@ -0,0 +1,395 @@
+                sorted_variables.clear();
+            }
+        }
+    }
+
+    auto& lookup_block = ultra_circuit_constructor.blocks.lookup;
+    auto lookup_gates = lookup_block.size();
+    bool lookup_gates_exists = lookup_gates > 0;
+    if (lookup_gates_exists) {
+        std::vector<uint32_t> left_indexes;
+        std::vector<uint32_t> right_indexes;
+        std::vector<uint32_t> out_indexes;
+        for (size_t i = 0; i < lookup_gates; i++) {
+            if (lookup_block.q_lookup_type()[i] == FF(1)) {
+                uint32_t left_idx = lookup_block.w_l()[i];
+                uint32_t right_idx = lookup_block.w_r()[i];
+                uint32_t out_idx = lookup_block.w_o()[i];
+                left_indexes.emplace_back(left_idx);
+                right_indexes.emplace_back(right_idx);
+                out_indexes.emplace_back(out_idx);
+                std::vector<uint32_t> current_gate_variables = { left_idx, right_idx, out_idx };
+                for (const auto& variable : current_gate_variables) {
+                    variables_gate_counts[variable] += 1;
+                }
+                connect_all_variables_in_vector(current_gate_variables, zero_idx, constant_variable_indices, false);
+                auto q_2 = lookup_block.q_2()[i];
+                auto q_m = lookup_block.q_m()[i];
+                auto q_c = lookup_block.q_c()[i];
+                if (q_2 == 0 && q_m == 0 && q_c == 0) {
+                    // this is the last gate in the in this blocks of lookup, so we can connect variables from
+                    //  left_indexes vectors, right_indexes vector, out_indexes vector
+                    connect_all_variables_in_vector(left_indexes, zero_idx, constant_variable_indices, false);
+                    connect_all_variables_in_vector(right_indexes, zero_idx, constant_variable_indices, false);
+                    connect_all_variables_in_vector(out_indexes, zero_idx, constant_variable_indices, false);
+                    // now we can erase all these vector to begin with new lookup accumulators
+                    left_indexes.clear();
+                    right_indexes.clear();
+                    out_indexes.clear();
+                }
+            }
+        }
+    }
+}
+
+template <typename FF>
+bool Graph_<FF>::check_is_not_constant_variable(const uint32_t& variable_index,
+                                                const std::map<FF, uint32_t>& constant_variable_indices)
+{
+    // this method checks that variable is not constant, so it's not in constant_variable_indices from ultra circuit
+    // builder
+    bool is_not_constant = true;
+    for (const auto& pair : constant_variable_indices) {
+        if (pair.second == variable_index) {
+            is_not_constant = false;
+            break;
+        }
+    }
+    return is_not_constant;
+}
+
+template <typename FF>
+void Graph_<FF>::connect_all_variables_in_vector(const std::vector<uint32_t>& variables_vector,
```

Comment:
Why are you still using zero index if you need to check that it's not a constant variable?